counsel.el

```diff
@@ -2245,18 +2245,50 @@ regex string."
 (ivy-set-occur 'counsel-ag 'counsel-ag-occur)
 (ivy-set-display-transformer 'counsel-ag 'counsel-git-grep-transformer)
 
+(defconst user-input-switch "-- ")
+(defconst user-input-switch-length (length user-input-switch))
+
+(defun parse-user-input (string)
+  "Parse user input in STRING into a pair of two strings: (query extra-ag-args).
+Where query is the search term and extra-ag-args are any additional switches."
+  (let ((query string)
+        (extra-ag-args ""))
+    (when (string-prefix-p "-" string)
+      (let ((index (string-match user-input-switch string)))
+       (when index
+          (setq query (substring string (+ user-input-switch-length index)))
+          (setq extra-ag-args (substring string 0 index)))))
+    (cons query extra-ag-args)))
+
+(defun format-counsel-ag-command (extra-ag-args term)
+  "Format counsel-ag-command as a string.
+EXTRA-AG-ARGS is a string of the additional arguments.
+TERM is the search string."
+  (let ((args-end (string-match " -- " extra-ag-args))
+        (file ""))
+    (when args-end
+      (setq file (substring-no-properties extra-ag-args (+ args-end 3)))
+      (setq extra-ag-args (substring-no-properties extra-ag-args 0 args-end)))
+    (format counsel-ag-command
+            (concat extra-ag-args
+                    term
+                    file))))
+
 (defun counsel-ag-function (string)
-  "Grep in the current directory for STRING using BASE-CMD.
-If non-nil, append EXTRA-AG-ARGS to BASE-CMD."
-  (if (< (length string) 3)
-      (counsel-more-chars 3)
-    (let ((default-directory (ivy-state-directory ivy-last))
-          (regex (counsel-unquote-regex-parens
-                  (setq ivy--old-re
-                        (ivy--regex string)))))
-      (counsel--async-command (format counsel-ag-command
-                                      (shell-quote-argument regex)))
-      nil)))
+  "Grep in the current directory for STRING."
+  (let ((parsed (parse-user-input string)))
+    (let ((query (car parsed))
+          (extra-ag-args (cdr parsed)))
```

Comment:
You can use a single `let*` instead of nested `let`.

Reply:
I though that since `query` and `extra-ag-args` don't depend on each other, only on `parsed` it'd make the relationship clearer. I can switch it to `let*` if reducing visual nesting makes it more readable.

Reply:
BTW, don't take my explanations as objections to make the changes. They are just additional information to inform the decision, but ultimately I trust that you are more familiar with the project and have a better feeling of how things work and affect it in the long term.
I other words don't hold back the *bike-shedding* 😄